[merged] WIP: Add support for mounting mqueue

mrunalp opened the pull request:
Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

mrunalp commented:
I tried adding support for mounting mqueue but saw some strange behavior most likely because of mount propagation.

```
[mrunal@dhcp-16-129 bubblewrap]$  bwrap  --unshare-pid  --unshare-ipc --bind ~/busybox / --proc /proc --dev /dev --mqueue /dev/mqueue  /bin/sh
/ $ cat /proc/self/mountinfo 
250 249 253:2 /mrunal/busybox / rw,nosuid,nodev,relatime master:34 - ext4 /dev/mapper/fedora_dhcp--16--129-home rw,seclabel,data=ordered
251 250 0:55 / /proc rw,nosuid,nodev,noexec,relatime - proc proc rw
252 251 0:55 /sys /proc/sys ro,nosuid,nodev,noexec,relatime - proc proc rw
253 251 0:55 /sysrq-trigger /proc/sysrq-trigger ro,nosuid,nodev,noexec,relatime - proc proc rw
254 251 0:55 /irq /proc/irq ro,nosuid,nodev,noexec,relatime - proc proc rw
255 251 0:55 /bus /proc/bus ro,nosuid,nodev,noexec,relatime - proc proc rw
256 250 0:62 / /dev rw,nosuid,nodev,noexec,relatime - tmpfs tmpfs rw,seclabel,mode=755,uid=1000,gid=1000
257 256 0:6 /null /dev/null rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
258 256 0:6 /zero /dev/zero rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
259 256 0:6 /full /dev/full rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
260 256 0:6 /random /dev/random rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
261 256 0:6 /urandom /dev/urandom rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
262 256 0:6 /tty /dev/tty rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
263 256 0:63 / /dev/pts rw,nosuid,noexec,relatime - devpts devpts rw,seclabel,mode=620,ptmxmode=666
264 256 0:15 /6 /dev/console rw,nosuid,noexec,relatime master:4 - devpts devpts rw,seclabel,gid=5,mode=620,ptmxmode=000
265 256 0:64 / /dev/mqueue rw,nosuid,nodev,noexec,relatime - tmpfs tmpfs rw,seclabel,mode=755,uid=1000,gid=1000
266 265 0:6 /null /dev/mqueue/null rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
267 265 0:6 /zero /dev/mqueue/zero rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
268 265 0:6 /full /dev/mqueue/full rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
269 265 0:6 /random /dev/mqueue/random rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
270 265 0:6 /urandom /dev/mqueue/urandom rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
271 265 0:6 /tty /dev/mqueue/tty rw,nosuid master:2 - devtmpfs devtmpfs rw,seclabel,size=8155136k,nr_inodes=2038784,mode=755
272 265 0:65 / /dev/mqueue/pts rw,nosuid,noexec,relatime - devpts devpts rw,seclabel,mode=620,ptmxmode=666
273 265 0:15 /6 /dev/mqueue/console rw,nosuid,noexec,relatime master:4 - devpts devpts rw,seclabel,gid=5,mode=620,ptmxmode=000
/ $ ls -l /dev/mqueue/
total 0
crw--w----    1 1000     nobody    136,   6 May  5 23:45 console
crw-rw-rw-    1 nobody   nobody      1,   7 Apr 27 18:46 full
crw-rw-rw-    1 nobody   nobody      1,   3 Apr 27 18:46 null
lrwxrwxrwx    1 1000     1000             8 May  5 23:45 ptmx -> pts/ptmx
drwxr-xr-x    2 nobody   nobody           0 May  5 23:45 pts
crw-rw-rw-    1 nobody   nobody      1,   8 Apr 27 18:46 random
drwxr-xr-x    2 1000     1000            40 May  5 23:45 shm
lrwxrwxrwx    1 1000     1000            15 May  5 23:45 stderr -> /proc/self/fd/2
lrwxrwxrwx    1 1000     1000            15 May  5 23:45 stdin -> /proc/self/fd/0
lrwxrwxrwx    1 1000     1000            15 May  5 23:45 stdout -> /proc/self/fd/1
crw-rw-rw-    1 nobody   nobody      5,   0 May  5 23:22 tty
crw-rw-rw-    1 nobody   nobody      1,   9 Apr 27 18:46 urandom
crw-rw-rw-    1 nobody   nobody      1,   5 Apr 27 18:46 zero
```

mrunalp commented:
@alexlarsson Any thoughts?

rhatdan commented:
I am not sure you need all of those mount flags.

In my system mqueue looks like the following.

mqueue on /dev/mqueue type mqueue (rw,relatime,seclabel)

Strange part is it looks like /dev/mqueue is being treated like it is /dev?

mrunalp commented:
@rhatdan I took out the flags and it makes no difference. I think we are probably leaking mounts from parent namespace. (We do leave the new root as MS_SLAVE to root namespace). 
Also this snippet from strace suggest so:
```
[pid  4544] close(5)                    = 0
[pid  4544] creat("/newroot/dev/mqueue/zero", 0666) = 5
[pid  4544] close(5)                    = 0
[pid  4544] mount("/oldroot/dev/zero", "/newroot/dev/mqueue/zero", NULL, MS_MGC_VAL|MS_BIND|MS_REC, NULL) = 0
[pid  4544] openat(3, "self/mountinfo", O_RDONLY|O_CLOEXEC) = 5
[pid  4544] read(5, "157 249 253:0 / /oldroot rw,rela"..., 4079) = 4074
[pid  4544] read(5, "244 2", 5)         = 5
[pid  4544] read(5, "43 8:1 / /oldroot/boot/efi rw,re"..., 4080) = 2408
[pid  4544] read(5, "", 1672)           = 0
```

alexlarsson commented on bubblewrap.c:
```diff
@@ -970,6 +988,17 @@ parse_args (int *argcp,
+          if (argc < 2)
+            die ("--mqueue takes an argument");
+
+          op = setup_op_new (SETUP_MOUNT_DEV);
```
s/SETUP_MOUNT_DEV/SETUP_MOUNT_MQUEUE/

mrunalp replied:
Ahh, thanks :) I should have looked closer!!!

alexlarsson commented:
So, i fixed the SETUP_MOUNT_DEV and removed the flags, but unprivileged i get:
```
5311  mount("mqueue", "/newroot/dev/mqueue", "mqueue", 0, NULL) = -1 EPERM (Operation not permitted)
```
and setuid i get:
```
5244  mount("mqueue", "/newroot/dev/mqueue", "mqueue", 0, NULL) = -1 EBUSY (Device or resource busy)
```

mrunalp commented:
I have rebased and fixed up the PR. It works for me though. Maybe the kernel version you have is causing issues? We have seen a report with similar issues in runc with ubuntu 16.04 on 4.4 kernel.

```
[mrunal@dhcp-16-129 bubblewrap]$ getcap `which bwrap`
[mrunal@dhcp-16-129 bubblewrap]$  bwrap  --unshare-pid  --unshare-ipc --bind ~/busybox / --proc /proc --mqueue /dev/mqueue  /bin/sh
/ $ cat /proc/self/mountinfo 
250 249 253:2 /mrunal/busybox / rw,nosuid,nodev,relatime master:34 - ext4 /dev/mapper/fedora_dhcp--16--129-home rw,seclabel,data=ordered
251 250 0:55 / /proc rw,nosuid,nodev,noexec,relatime - proc proc rw
252 251 0:55 /sys /proc/sys ro,nosuid,nodev,noexec,relatime - proc proc rw
253 251 0:55 /sysrq-trigger /proc/sysrq-trigger ro,nosuid,nodev,noexec,relatime - proc proc rw
254 251 0:55 /irq /proc/irq ro,nosuid,nodev,noexec,relatime - proc proc rw
255 251 0:55 /bus /proc/bus ro,nosuid,nodev,noexec,relatime - proc proc rw
256 250 0:54 / /dev/mqueue rw,nodev,relatime - mqueue mqueue rw,seclabel
/ $ exit
[mrunal@dhcp-16-129 bubblewrap]$ uname -a
Linux dhcp-16-129.sjc.redhat.com 4.6.0-0.rc4.git3.1.fc25.x86_64 #1 SMP Fri Apr 22 16:16:10 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

cgwalters commented:
Works for me on `3.10.0-327.13.1.el7.x86_64` as setuid.

@cgwalters-bot r+ 54497e0
